### PR TITLE
Properly handle objects while changing terrain

### PR DESCRIPTION
**Pull request by ihhub**
- remove objects that do not belong to the new terrain
- update castles' basements according to the new terrain

relates to #6845

Things to test:
- erasure tool works as intended
- placing water over objects that cannot be placed on water (and vice versa) results in their removal
- town basements should be updated according to new terrain
- general object placement on different terrain types

### CODE REVIEW

**Comment by Branikolog** [comment]
Hi, @ihhub 

Multiple assertions occur, when I've tried to place any terrain over the existing, after I've placed several map objects. 
image

**Comment by ihhub** [comment]
Hi @Branikolog , I've fixed the crashes.

**Comment by Branikolog** [comment]
@ihhub 

One object does not seem to be affected by terrain change:

https://github.com/ihhub/fheroes2/assets/55348946/65727b8a-6e39-4776-9b17-1f6f3f483a18

**Comment by ihhub** [comment]
Hi @Branikolog , river delta is terrain independent by design.

**Review by Branikolog — APPROVED** [review]
So no other problems occurred. 👍

File: src/fheroes2/editor/editor_interface.cpp [hunk 1]
```diff
@@ -395,6 +392,217 @@ namespace
...
+                return false;
+            }
+        }
+        else if ( groupType == Maps::ObjectGroup::LANDSCAPE_TREES ) {
+            const auto & objectInfo = Maps::getObjectInfo( groupType, objectType );
+            if ( !checkConditionForUsedTiles( objectInfo, tilePos, []( const Maps::Tiles & tileToCheck ) { return !tileToCheck.isWater(); } ) ) {
+                errorMessage = _( "Trees cannot be placed on water." );
+                return false;
+            }
+        }
```
Comment (Districh-ru) [inline src/fheroes2/editor/editor_interface.cpp:446]: Can we combine these mostly identical code, in example:
```cpp
        if ( groupType != Maps::ObjectGroup::ADVENTURE_WATER && groupType != Maps::ObjectGroup::LANDSCAPE_WATER ) {
            const auto & objectInfo = Maps::getObjectInfo( groupType, objectType );
            if ( !checkConditionForUsedTiles( objectInfo, tilePos, []( const Maps::Tiles & tileToCheck ) { return !tileToCheck.isWater(); } ) ) {
                errorMessage = _( "%{objectGroup} cannot be placed on water." );
                StringReplace( text, "%{objectGroup}", getObjectGroupName( groupType ) );
                return false;
            }
        }
```

and introduce `getObjectGroupName()` function to return group name?

File: src/fheroes2/editor/editor_interface.cpp [hunk 2]
```diff
@@ -395,6 +392,217 @@ namespace
...
+
+            if ( !isObjectPlacementAllowed( townObjectInfo, tilePos ) ) {
+                errorMessage = _( "Objects cannot be placed outside the map." );
+                return false;
+            }
+
+            if ( !isObjectPlacementAllowed( basementObjectInfo, tilePos ) ) {
+                errorMessage = _( "Objects cannot be placed outside the map." );
+                return false;
+            }
```
Comment (Districh-ru) [inline src/fheroes2/editor/editor_interface.cpp:589]: Can we join these two conditions?
`if ( !isObjectPlacementAllowed( townObjectInfo, tilePos ) || !isObjectPlacementAllowed( basementObjectInfo, tilePos ) )` ...

**Review by Districh-ru — COMMENTED** [review]
Hi, @ihhub, I left two comments. Could you please check them when you have time?

File: src/fheroes2/editor/editor_interface.cpp (hunk 2 above)
Comment (ihhub) [reply src/fheroes2/editor/editor_interface.cpp:589]: I believe it is going to be very lengthy and ugly to read :(

Comment (ihhub) [reply src/fheroes2/editor/editor_interface.cpp:553]: Actually, you are right. I was looking at the different place.

**Pull request merged**